File: source/ts/src/core/yoga.ts
import { normalize360 } from '../utils/angle';
import { YOGA_SPAN } from '../utils/constants';
import type { YogaInfo } from '../types/elements';

export function computeYogaFromLongitudes(
  siderealMoon: number,
  siderealSun: number,
  name: string,
): YogaInfo {
  const angle = normalize360(siderealSun + siderealMoon);
  const index = Math.floor(angle / YOGA_SPAN);
  const elapsed = angle - index * YOGA_SPAN;
  const completionPercentage = (elapsed / YOGA_SPAN) * 100;
  return {
    index,
    name,
    completionPercentage: Math.round(completionPercentage * 100) / 100,
    endTime: null,
  };
}

export function getYogaIndexAtTime(
  date: Date,
  getCachedMoon: (d: Date) => number,
  getCachedSun: (d: Date) => number,
): number {
  const angle = normalize360(getCachedSun(date) + getCachedMoon(date));
  return Math.floor(angle / YOGA_SPAN);
}

export function getYogaIndex(siderealMoon: number, siderealSun: number): number {
  return Math.floor(normalize360(siderealSun + siderealMoon) / YOGA_SPAN);
}

File: source/go/internal/core/yoga.go
package core

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

func ComputeYogaFromLongitudes(siderealMoon, siderealSun float64, name string) types.YogaInfo {
	angle := utils.Normalize360(siderealSun + siderealMoon)
	index := int(math.Floor(angle / utils.YogaSpan))
	elapsed := angle - float64(float64(index)*utils.YogaSpan) // anti-FMA barrier
	completionPercentage := (elapsed / utils.YogaSpan) * 100
	return types.YogaInfo{
		Index:                index,
		Name:                 name,
		CompletionPercentage: jsnum.Round(completionPercentage*100) / 100,
		EndTime:              nil,
	}
}

func GetYogaIndexAtTime(ms int64, getCachedMoon, getCachedSun LongitudeAt) int {
	angle := utils.Normalize360(getCachedSun(ms) + getCachedMoon(ms))
	return int(math.Floor(angle / utils.YogaSpan))
}

func GetYogaIndex(siderealMoon, siderealSun float64) int {
	return int(math.Floor(utils.Normalize360(siderealSun+siderealMoon) / utils.YogaSpan))
}
